### PR TITLE
🎨  logging

### PR DESCRIPTION
**Pull request by kirrg001**
refs #7116 

First iteration of logging and error refactoring/usage. You can follow each commit to see what this PR already contains.

Please read through all `TODO'S` here:
https://github.com/TryGhost/Ghost/issues/7116

---
### Things to know

Logging errors/requests happens automatically for each HTTP request.
If an error happens outside of an HTTP call, `logging.error` can be used manual.

Two more PR's will follow: 
- one for changing how we create errors + add error inheritance
- one for making final changes to logging

---
### logging config

With this PR we offer for now 3 attributes to set inside the config.

(this is the default one!)

```
"logging": {
    "level": "info",
    "rotation": false,
    "transports": ["stdout"]
}
```

_level_ defines which outputs should be displayed.
_rotation_ enable log file rotation (i can think of changing this to rotation: 1000 \* 60 \* 24)
_transports_ an array to define the transports (right now we offer file and stdout) - if we offer to add custom logger adapters, file transport can be defined for example as GELF stream and not as default write into file.

---
### 11th commit explained (https://github.com/TryGhost/Ghost/pull/7431/commits/cb990a3d8bd022beb5afd76ce6e63b58b7b09440)

> https://github.com/TryGhost/Ghost/pull/7431/commits/cb990a3d8bd022beb5afd76ce6e63b58b7b09440#diff-e1c52a506c4dfb4202439ef6db5f2de0L325

`canThis` will return NoPermissionError. If `findOne` returns null, it will still return NoPermissionError.

> https://github.com/TryGhost/Ghost/pull/7431/commits/cb990a3d8bd022beb5afd76ce6e63b58b7b09440#diff-76bafa2cda0d4f737b1269c2dd682497R20

Looks right now a bit ugly this change, but will look better in #7477 .

> https://github.com/TryGhost/Ghost/pull/7431/commits/cb990a3d8bd022beb5afd76ce6e63b58b7b09440#diff-e2843b3b16d9e7a6b39f8592f2277b3dL58

Right now there is no 404 page handler (except of this https://github.com/TryGhost/Ghost/pull/7431/commits/e108ecd6889f8871f2a283836f3c70acd97b8ff2#diff-7a5f479230bbe115458d26c3c7026164R217). We can add one, but this would just wrap the `NotFoundError` and calls `next`. In future thinking: if i18n translation exists as default configuration of an Error (`NotFoundError: function() { this.i18nKey: 'page-not-found'})`), then we don't need a PageNotFound Handler.

> https://github.com/TryGhost/Ghost/pull/7431/commits/cb990a3d8bd022beb5afd76ce6e63b58b7b09440#diff-10376cebb233c43926136766fbebfbd6R70

Tested, works. Easy way to forward error to browser.

> https://github.com/TryGhost/Ghost/pull/7431/commits/cb990a3d8bd022beb5afd76ce6e63b58b7b09440#diff-0448a77db707d046e3c69f1274d75c42R47

No need to log, happens in HTTP request.
Help/Context is not forwarded to Client, but get's logged into server logs!

> https://github.com/TryGhost/Ghost/pull/7431/commits/cb990a3d8bd022beb5afd76ce6e63b58b7b09440#diff-430e6e1bb13856cddb780c0847022234R103

I am not 100% sure about this change, but seems to me it makes more sense to trigger the error-handler then just throwing an error?

**The rest is in my opinion is straight forwarded replacement by the new `logging` component.**

---
### First impression

The important fact is: file logging and stdout is handled different.
Why?Just printing json to stdout is hard to read.

screen shot 2016-09-23 at 15 51 53

lol

screen shot 2016-09-23 at 15 54 06

screen shot 2016-09-23 at 15 52 12

screen shot 2016-09-23 at 15 52 40

### CODE REVIEW

File: core/server/models/role.js [hunk 1]
```diff
@@ -51,13 +51,15 @@ Role = ghostBookshelf.Model.extend({
...
             origArgs = _.toArray(arguments).slice(1);
+
             // Get the actual role model
+            // @TODO: test
             return this.findOne({id: roleModelOrId, status: 'all'}).then(function then(foundRoleModel) {
                 // Build up the original args but substitute with actual model
                 var newArgs = [foundRoleModel].concat(origArgs);
 
                 return self.permissible.apply(self, newArgs);
-            }, errors.logAndThrowError);
```
Comment (ErisDS) [inline core/server/models/role.js]: This changes behaviour, is it intentional / tested?

File: core/server/models/settings.js [hunk 2]
```diff
@@ -138,14 +139,14 @@ Settings = ghostBookshelf.Model.extend({
                     return setting.save(saveData, options);
                 }
 
-                return Promise.reject(new errors.NotFoundError(i18n.t('errors.models.settings.unableToFindSetting', {key: item.key})));
-            }, errors.logAndThrowError);
```
Comment (ErisDS) [inline core/server/models/settings.js]: Again this is a behaviour change

File: core/server/storage/local-file-store.js [hunk 3]
```diff
@@ -100,7 +100,19 @@ LocalFileStore.prototype.serve = function (options) {
...
+                    if (err.statusCode === 404) {
+                        return next(new errors.NotFoundError({message: i18n.t('errors.errors.pageNotFound')}));
+                    }
+
+                    return next(new errors.GhostError({err: err}));
+                }
+
+                next();
+            });
+        };
```
Comment (ErisDS) [inline core/server/storage/local-file-store.js]: Not sure what this change is for?

File: core/server/api/authentication.js [hunk 4]
```diff
@@ -42,7 +43,7 @@ function assertSetupCompleted(status) {
                 notCompleted = i18n.t('errors.api.authentication.setupMustBeCompleted');
 
             function throwReason(reason) {
-                throw new errors.NoPermissionError(reason);
+                throw new errors.NoPermissionError({message: reason});
```
Comment (ErisDS) [inline core/server/api/authentication.js]: I don't understand the reason for these changes. There are a lot of them and they make this PR incredibly difficult to review. If we are going to change our error signature to be different, I think this should probably be a  separate PR?

File: core/server/logging/PrettyStream.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
+// jscs:disable
```
Comment (ErisDS) [inline core/server/logging/PrettyStream.js]: I don't quite understand this file. It seems to do an awful lot of work, was there not already a library that could do this?

File: core/server/middleware/labs.js [hunk 6]
```diff
@@ -7,7 +7,7 @@ labs = {
         if (labsUtil.isSet('subscribers') === true) {
             return next();
         } else {
-            return errors.handleAPIError(new errors.NotFoundError(), req, res, next);
+            return next(new errors.NotFoundError());
```
Comment (ErisDS) [inline core/server/middleware/labs.js]: I didn't find the piece where the API errors are implemented now?

File: core/server/middleware/theme-handler.js [hunk 7]
```diff
@@ -78,9 +79,6 @@ themeHandler = {
 
         blogApp.engine('hbs', hbs.express3(hbsOptions));
 
-        // Update user error template
-        errors.updateActiveTheme(activeTheme);
```
Comment (ErisDS) [inline core/server/middleware/theme-handler.js]: Where is the other half of this - where this is reimplemented?

File: core/server/models/user.js [hunk 8]
```diff
@@ -456,13 +459,15 @@ User = ghostBookshelf.Model.extend({
...
             origArgs = _.toArray(arguments).slice(1);
+
             // Get the actual user model
+            // @TODO: test
             return this.findOne({id: userModelOrId, status: 'all'}, {include: ['roles']}).then(function then(foundUserModel) {
                 // Build up the original args but substitute with actual model
                 var newArgs = [foundUserModel].concat(origArgs);
 
                 return self.permissible.apply(self, newArgs);
-            }, errors.logAndThrowError);
```
Comment (ErisDS) [inline core/server/models/user.js]: Another behaviour change.

File: core/server/models/user.js [hunk 9]
```diff
@@ -536,53 +541,52 @@ User = ghostBookshelf.Model.extend({
...
                             // If we get a validation or other error during this save, catch it and log it, but don't
                             // cause a login error because of it. The user validation is not important here.
-                            errors.logError(
-                                error,
-                                i18n.t('errors.models.user.userUpdateError.context'),
-                                i18n.t('errors.models.user.userUpdateError.help')
-                            );
                             return user;
                         });
-                }, errors.logAndThrowError);
```
Comment (ErisDS) [inline core/server/models/user.js]: Again, all these small changes matter a lot IMO? They are hidden away by all the changes to the error signature. I really don't think that this should be in the same PR.

**Review by ErisDS — CHANGES_REQUESTED** [review]
I have left some comments on some pieces of code. Overall I have no visibility on what the new logging code is really looking like just from this PR. I will also checkout the code.

I also cannot see where request logging has been implemented?

File: core/server/models/role.js (hunk 1 above)
Comment (kirrg001) [inline core/server/models/role.js]: I thought about and I don't think it's a behaviour change, but tell me if you still think it is after explaining why.


`errors.logAndThrowError` get's called when `this.findOne` throws an Error.
`errors.logAndThrowError` logs and throws the error (again)

This function call **always** happens in a HTTP request, so logging happens now automatically through - no need to log it twice (https://github.com/kirrg001/Ghost/blob/logging/core/server/middleware/index.js#L88)

And i don't see any reason why to throw the error?

File: core/server/storage/local-file-store.js (hunk 3 above)
Comment (kirrg001) [inline core/server/storage/local-file-store.js]: I pushed this change as separate commit, see commit message for reasoning.
https://github.com/TryGhost/Ghost/pull/7431/commits/41747750253f2867a6d9d8f1a33e24391525bfd8

i hope that helps.

File: core/server/api/authentication.js (hunk 4 above)
Comment (kirrg001) [inline core/server/api/authentication.js]: separate PR about how we create errors will come tomorrow 👍

File: core/server/logging/PrettyStream.js (hunk 5 above)
Comment (kirrg001) [inline core/server/logging/PrettyStream.js]: Wasn't lot of work, maybe 30mins. I just wrote the way we log, but not how pretty streaming works. This is a general approach and copied. 

There was no satisfying stdout module.

File: core/server/middleware/labs.js (hunk 6 above)
Comment (kirrg001) [inline core/server/middleware/labs.js]: i hope that helps:
https://github.com/TryGhost/Ghost/pull/7431/commits/da4ce0a9e37339c4494b33266ea23b8142a8ec47

File: core/server/middleware/theme-handler.js (hunk 7 above)
Comment (kirrg001) [inline core/server/middleware/theme-handler.js]: i hope that helps
https://github.com/TryGhost/Ghost/pull/7431/commits/114cc12feaf32b9214f515747137a7dbcdde62ea

File: core/server/middleware/labs.js (hunk 6 above)
Comment (kirrg001) [inline core/server/middleware/labs.js]: plus: https://github.com/kirrg001/Ghost/blob/logging/core/server/middleware/index.js#L216

File: core/server/middleware/error-handler.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * @TODO: jsonapi errors format
+ */
+_private.handleJSONResponse = function handleJSONResponse(err, req, res) {
+    return function handleJSONResponse() {
+        res.json({
+            errors: [{
+                message: err.message,
+                errorType: err.errorType
```
Comment (kirrg001) [inline core/server/middleware/error-handler.js]: `TODO`: return `errorDetails` and look if something else is missing for Ghost-Admin

Comment (kirrg001) [inline core/server/middleware/error-handler.js]: https://github.com/TryGhost/Ghost/pull/7431/commits/af8736f6cf82294f41ce75182a4a7331bda484f5

File: core/server/middleware/theme-handler.js [hunk 11]
```diff
@@ -99,15 +100,14 @@ themeHandler = {
                 // Change theme
                 if (!config.get('paths').availableThemes.hasOwnProperty(activeTheme.value)) {
                     if (!res.isAdmin) {
-                        // Throw an error if the theme is not available, but not on the admin UI
-                        return errors.throwError(i18n.t('errors.middleware.themehandler.missingTheme', {theme: activeTheme.value}));
+                        return next(new errors.NotFoundError(i18n.t('errors.middleware.themehandler.missingTheme', {theme: activeTheme.value})));
```
Comment (ErisDS) [inline core/server/middleware/theme-handler.js]: I think this is a 500-level error, rather than a 404? 4** is a client error, e.g. the client looking for something that doesn't exist, but this is a server error where the server is missing something vital to its operation?

File: core/server/middleware/index.js [hunk 12]
```diff
@@ -79,14 +78,17 @@ setupMiddleware = function setupMiddleware(blogApp) {
...
-        }
-    }
+    /**
+     * request logging
+     * @TODO: add error to output
+     */
+    blogApp.use(function expressLogging(req, res, next) {
+        res.once('finish', function () {
+            logging.request({req: req, res: res});
+        });
```
Comment (ErisDS) [inline core/server/middleware/index.js]: Have done some testing, we're not recording a duration. I think this needs improving, if not right now, in a later iteration.

**Review by ErisDS — CHANGES_REQUESTED** [review]
Most of my comments inline are minor or suggestions for TODOs for later iterations.

I have a few major concerns:
- The file-based rotation logging seems to be configured incorrectly. It only seems to log for 1 minute before rotating, and only keeps 3 minutes of history!
- Boot up is not generating any logging, so there is nothing in the logs until the first request
- The stdout pretty printer seems super weird to me, it's using custom code and the output is not like anything I've ever seen before. 

I'd like to know what the reference material for the output is, and to do another review of existing implementations. I think we should pick a reference to emulate rather than inventing our own? E.g. emulate morgan to start with.

File: core/server/errors/index.js [hunk 13]
```diff
@@ -365,7 +358,8 @@ errors = {
         }
 
         // Are we admin? If so, don't worry about the user template
-        if ((res.isAdmin && req.user && req.user.id) || userErrorTemplateExists === true) {
+        if ((res.isAdmin && req.user && req.user.id) ||
+            (config.get('paths').availableThemes[req.app.get('activeTheme')] || {}).hasOwnProperty('error.hbs') === true) {
```
Comment (kirrg001) [inline core/server/errors/index.js]: 🎉

File: core/server/middleware/theme-handler.js (hunk 11 above)
Comment (kirrg001) [inline core/server/middleware/theme-handler.js]: yeah lets use internal server error 500 👍

Comment (kirrg001) [inline core/server/middleware/theme-handler.js]: https://github.com/TryGhost/Ghost/pull/7431/commits/37bc2d961ad8243205d3310d2ea259772b3e7d38

File: core/server/middleware/error-handler.js [hunk 14]
```diff
@@ -0,0 +1,109 @@
+var _ = require('lodash'),
```
Comment (kirrg001) [inline core/server/middleware/error-handler.js]: yes totally agree 👍 
as you said, this is just the first refactoring of having cleaner code and especially cleaner handling!
later it's easy to just split both error handlers.

File: core/server/config/env/config.production.json [hunk 15]
```diff
@@ -8,5 +8,10 @@
     },
     "paths": {
         "contentPath": "content/"
+    },
+    "logging": {
+        "level": "info",
+        "rotation": true,
+        "transports": ["file"]
```
Comment (kirrg001) [inline core/server/config/env/config.production.json]: it's not planned to have multiple log files.
in HTTP world, error and request belong together in 99%. 
especially when developers (or we) want to collect logs into a centralised system, it's very helpful to have them together. so each log entry contains: the request, the response and an error if one occurred. This is one entry.

**Comment by kirrg001** [comment]
> Most of my comments inline are minor or suggestions for TODOs for later iterations.
I have a few major concerns:
The file-based rotation logging seems to be configured incorrectly. It only seems to log for 1 minute before rotating, and only keeps 3 minutes of history!

Yeah needs to be adapted!!!This is a left over from testing. I think weekly makes sense, will change now.

> Boot up is not generating any logging, so there is nothing in the logs until the first request

Are you talking about production mode?It only streams into file.
Development prints the normal print outs.


> The stdout pretty printer seems super weird to me, it's using custom code and the output is not like anything I've ever seen before.

Yeah it's a custom Ghost pretty printer for now. The whole pretty streamer can be changed to work generic, but not needed for now. It can take a bunch of parameters and can behave dynamically in the future.

> I'd like to know what the reference material for the output is, and to do another review of existing implementations. I think we should pick a reference to emulate rather than inventing our own? E.g. emulate morgan to start with.

Do you mean the stdout prints?
1. the custom stdout are only used in dev! (production prints json into file)
2. all pretty stdout printers i tested so far where not good enough (not pretty enough)
3. I can list that research in the next iteration 👍

**Comment by ErisDS** [comment]
>> Most of my comments inline are minor or suggestions for TODOs for later iterations.
I have a few major concerns:
The file-based rotation logging seems to be configured incorrectly. It only seems to log for 1 minute before rotating, and only keeps 3 minutes of history!

> Yeah needs to be adapted!!!This is a left over from testing. I think weekly makes sense, will change now.

I think this should also be keeping at least 100 files and I think this configuration needs to be exposed somehow.

>> Boot up is not generating any logging, so there is nothing in the logs until the first request

> Are you talking about production mode?It only streams into file.
> Development prints the normal print outs.

We need to make sure the logs (no matter what mode) get some indication of server start and server stop. This is a TODO for later.


>> I'd like to know what the reference material for the output is, and to do another review of existing implementations. I think we should pick a reference to emulate rather than inventing our own? E.g. emulate morgan to start with.

> Do you mean the stdout prints?
1. the custom stdout are only used in dev! (production prints json into file)
2. all pretty stdout printers i tested so far where not good enough (not pretty enough)
3. I can list that research in the next iteration 👍 

Yes, I'm talking about what's output to stdout. It's only use in dev given the default config. It can be used in production if someone chooses that. I think we should emulate an existing good solution (like morgan) rather than have our own custom one. Yes we can revisit this later.

File: core/server/middleware/index.js (hunk 12 above)
Comment (ErisDS) [inline core/server/middleware/index.js]: In the #7116 issue the output form morgan looks like:

```
GET / 200 140.121 ms - -
GET /assets/js/jquery.fitvids.js?v=4d853bd599 200 17.077 ms - -
GET /assets/js/index.js?v=4d853bd599 200 16.385 ms - -
GET /assets/css/screen.css?v=4d853bd599 200 19.621 ms - -
```

This includes a duration, which is missing from the current code. Again, this can be a TODO for a later iteration, however I think we should use the same code for logging to express as morgan does.

File: core/server/config/env/config.production.json (hunk 15 above)
Comment (ErisDS) [inline core/server/config/env/config.production.json]: This limitation seems wrong to me and doesn't fit with what is written in the goals on the issue: 
> full logging which outputs to stdout/stderr as we do now (and better!)

If stdout and stderr are different streams, then it would make sense to be able to log them to different places. This is pretty standard across things like nginx and apache, and is also what is currently done on Ghost(Pro). I'm not sure if the existing implementation allows this to be done differently by using a different adapter?

File: core/server/middleware/index.js (hunk 12 above)
Comment (kirrg001) [inline core/server/middleware/index.js]: ah sure 👍 added TODO

File: core/server/config/env/config.production.json (hunk 15 above)
Comment (kirrg001) [inline core/server/config/env/config.production.json]: Custom logging adapters don't have a final design, but from what i can see now, custom adapters can override what they want. They will inherit from the npm logging base and then it's up to them.

We are not restricted to anything, we could print all requests with errors in one file and all requests without errors in to one. With this PR, all http requests with and without errors land in one file. I will add a `TODO` to the issue as well to reconsider what's best  👍

**Review by ErisDS — APPROVED** [review]

**Comment by ErisDS** [comment]
I think we've captured all the todos for this, and come up with a plan for how to spec out the final configuration (both defaults, and what can be changed). Going to get this merged to unblock other things, there will be more iterations 😁

File: core/server/config/env/config.production.json (hunk 15 above)
Comment (sebgie) [inline core/server/config/env/config.production.json]: It is possible to register multiple streams with bunyan that operate on different log levels or even different serializers. If we expose this functionality in the config we could have different log files (or even different systems) with different visibility.

Separate logs for errors and requests are quite common in HTTP servers (Apache, Nginx) to get a quick overview of what is going wrong when a request fails. On a high traffic site it's way easier to find what you are looking for when there is a separate error log.

I think that we can add that functionality at a later alpha but it would be very handy to have it available!

Comment (kirrg001) [inline core/server/config/env/config.production.json]: > Separate logs for errors and requests are quite common in HTTP servers (Apache, Nginx) to get a quick overview of what is going wrong when a request fails.

Yes i am aware of that. And i agree, having a separate error log file might be useful, which only contains requests with errors (or single errors outside of the HTTP scope). But in future thinking, in bigger software nobody looks at plain log files. People send their logs to centralised logging system. And with this tool they can do what they want with the JSON log files. (order, send mails when errors are critical, analyse user behaviour etc.).

> It is possible to register multiple streams with bunyan that operate on different log levels or even different serializers. If we expose this functionality in the config we could have different log files (or even different systems) with different visibility.

Yeah, we can register what we want, we just need to define what we need (or what people need). I will write up a more detailed expectation for the default logging component we will offer into #7431 . This will contain: logging config and default logging behaviour. I am happy about feedback then :)

File: core/server/middleware/error-handler.js (hunk 10 above)
Comment (ErisDS) [inline core/server/middleware/error-handler.js]: One other @TODO to log here - we've lost the logic that removes the stack trace for 404 errors and also for database errors on startup.
